src/cosim/serialization.cpp

```diff
@@ -0,0 +1,99 @@
+#include "cosim/serialization.hpp"
+
+#include <algorithm>
+#include <ios>
+#include <iterator>
+#include <utility>
+
+
+namespace
+{
+    // Visitor class which prints the value(s) contained in a leaf node to an
+    // output stream and returns whether anything was written.
+    struct leaf_output_visitor
+    {
+        leaf_output_visitor(std::ostream& out) : out_(out) { }
+
+        template<typename T>
+        bool operator()(T value) { out_ << value; return true; }
+
+        bool operator()(std::nullptr_t) { return false; }
+
+        bool operator()(bool value)
+        {
+            const auto flags = out_.flags();
+            out_ << std::boolalpha << value;
+            out_.flags(flags);
+            return true;
+        }
+
+        bool operator()(char value)
+        {
+            out_ << '\'' << value << '\'';
+            return true;
+        }
+
+        bool operator()(const std::string& str)
+        {
+            out_ << '"' << str << '"';
+            return true;
+        }
+
+        bool operator()(std::byte value)
+        {
+            const auto flags = out_.flags();
+            out_ << "0x" << std::hex << int(value);
+            out_.flags(flags);
+            return true;
+        }
```

Comment:
This is not visited from `std::vector<std:;byte>`?

Reply:
First of all, note that this visitor is only used in `operator<<`, which is just intended as a tool for developing and debugging libcosim itself. I even suspect I'll remove it before this goes into `master`.

That said, the answer to your question is no. I was worried that some FMUs might have a lot of data in their exported state, so printing it all out as a hex sequence would be counterproductive. Also, the actual serialisation format is a hidden implementation detail of each FMU, so I'm not sure it would give us much.

For `std::vector<std::byte>` it therefore just prints `<binary data, N bytes>` where `N` is the size of the blob. I'm happy to reconsider this if you think the functionality would be generally useful for something.

Reply:
Okay, I thought this would be used for actually serializing the entire states.